nexus/src/db/datastore/role.rs

```diff
@@ -134,13 +134,38 @@ impl DataStore {
         // exposed via an external API right now but someone could still put us
         // into some hurt by assigning loads of roles to someone and having that
         // person attempt to access anything.
-        dsl::role_assignment
-            .filter(dsl::identity_type.eq(identity_type))
-            .filter(dsl::identity_id.eq(identity_id))
-            .filter(dsl::resource_type.eq(resource_type.to_string()))
-            .filter(dsl::resource_id.eq(resource_id))
-            .select(RoleAssignment::as_select())
-            .load_async::<RoleAssignment>(self.pool_authorized(opctx).await?)
+
+        self.pool_authorized(opctx).await?
+            .transaction(move |conn| {
+                let mut role_assignments = dsl::role_assignment
+                    .filter(dsl::identity_type.eq(identity_type.clone()))
+                    .filter(dsl::identity_id.eq(identity_id))
+                    .filter(dsl::resource_type.eq(resource_type.to_string()))
+                    .filter(dsl::resource_id.eq(resource_id))
+                    .select(RoleAssignment::as_select())
+                    .load::<RoleAssignment>(conn)?;
+
+                // Return the roles that a silo user has from their group memberships
+                if identity_type == IdentityType::SiloUser {
+                    use db::schema::silo_group_membership;
+
+                    let mut group_role_assignments = dsl::role_assignment
+                        .filter(dsl::identity_type.eq(IdentityType::SiloGroup))
+                        .filter(dsl::identity_id.eq_any(
+                            silo_group_membership::dsl::silo_group_membership
```

Comment:
I think this will work okay.  It's a little different than what I'd suggested, which was:

```sql
SELECT r.* from role_assignment r, silo_group_membership m WHERE
    m.silo_user_id == $1 AND
    r.actor_type == 'silo_group' AND
    r.actor_id == m.silo_group_id AND
    r.resource_type == $2 AND
    r.resource_id = $3
```

This looks more like:

```sql
SELECT * from role_assignment WHERE
    r.identity_type == 'silo_group' AND
    r.identity_id IN (SELECT silo_group_id FROM silo_group_membership WHERE silo_user_id = $1) AND
    r.resource_type == $2 AND
    r.resource_id = $3
```

This uses a subquery instead of a join.  The main downside is that the database will materialize the full list of a user's groups and check that set against all the matching role assignments.  The join might also have to do that but I think the database would have more freedom to choose the best approach (e.g., if there were a large number of role assignments and few groups, it could choose instead to enumerate the groups and check each one for a matching role assignment).  I assume group membership will be small for the foreseeable future and this won't be a problem.